api/OBJECT-SIGNING.md

```diff
@@ -0,0 +1,63 @@
+# API Object Signing
+
+In the V2 API, in addition to using TLS for server/proxy authentication and encryption, the API uses object signing using the [auth.AuthContainer](./proto/auth/auth.proto) to provide end-to-end integrity of the message payloads.
+
+## Motivation
+
+The motivation for this is two-fold:
+
+- TLS Man-in-the-Middle proxy products and server-side load balancing and DDoS protection services do not work well with TLS client certificates and the restricted single server root CA certificate used in the V1 API.
+- The desire to be able to carry signed edge-node configuration through some non-TLS channel such as a USB memory stick.
+
+## Certificate management
+
+When TLS MiTM is in place the device will trust the proxy at the TLS level either because it is trusted by being in the standard Linux root certificate list, or because it was specified in the `proxyCertPEM` field in the device configuration API. Thus TLS still provides server authentication and privacy plus integrity.
+
+But since there is no TLS client authentication the `AuthContainer` message is used to sign the payloads from the device to the controller (using standard ECDA signing operations (implemented in ecdsa golang package alternatively the Sign function from github.com/google/go-tpm/tpm2 when the device private key is in a TPM).
```

Comment:
I'd disagree with the statement that we are replicating part of TLS. We are doing PK signatures. TLS doesn't do that. TLS does 100 other things.

Reply:
> TLS does 100 other things

Very good point. But we are replicating the two most important parts, aren't we? Message validation (signing) and message protection (encryption). That is the (use-case) basis of all TLS. The USB case certainly covers protection of data in transit.

Reply:
No, we are merely augmenting the integrity protection (call it signing or message authentication) between the TLS endpoints with object signing between the object endpoints. Plus we are adding message origin authentication for device to controller instead of using TLS client certificates.